plugins/permission-backend/src/service/router.ts

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  errorHandler,
+  PluginEndpointDiscovery,
+} from '@backstage/backend-common';
+import express, { Request, Response } from 'express';
+import Router from 'express-promise-router';
+import { Logger } from 'winston';
+import {
+  BackstageIdentity,
+  IdentityClient,
+} from '@backstage/plugin-auth-backend';
+import { ConflictError } from '@backstage/errors';
+import {
+  AuthorizeResult,
+  AuthorizeResponse,
+  AuthorizeRequest,
+  Identified,
+} from '@backstage/plugin-permission-common';
+import { PermissionPolicy } from '@backstage/plugin-permission-node';
+import { PermissionIntegrationClient } from './PermissionIntegrationClient';
+
+/**
+ * Options required when constructing a new {@link express#Router} using
+ * {@link createRouter}.
+ *
+ * @public
+ */
+export interface RouterOptions {
+  logger: Logger;
+  discovery: PluginEndpointDiscovery;
+  policy: PermissionPolicy;
+}
+
+const handleRequest = async (
+  { id, resourceRef, ...request }: Identified<AuthorizeRequest>,
+  user: BackstageIdentity | undefined,
+  policy: PermissionPolicy,
+  permissionIntegrationClient: PermissionIntegrationClient,
+  authHeader?: string,
+): Promise<Identified<AuthorizeResponse>> => {
+  const response = await policy.handle(request, user);
+
+  if (response.result === AuthorizeResult.CONDITIONAL) {
+    // Sanity check that any resource provided matches the one expected by the permission
+    if (request.permission.resourceType !== response.conditions.resourceType) {
+      throw new ConflictError(
```

Comment:
Would `InputError` be a better fit here? Seems more like a misconfigured policy rather than a conflict in server state 🤔

Reply:
Mmm, I’m a bit conflicted (ha) myself. I do see that it’s not quite a “conflict”, but it’s also not an error in the input from the perspective of the caller, since they don’t necessarily have control over the policy.

Reply:
A 500 could also be appropriate here! 4xx errors could go unnoticed and a misconfigured policy should probably call more attention than that

Reply:
Yeah, good call! Switched over in 726f84a.